File: java/service/iothub-service-sdk/src/main/java/com/microsoft/azure/iot/service/sdk/Receiver.java
/*
 * Copyright (c) Microsoft. All rights reserved.
 * Licensed under the MIT license. See LICENSE file in the project root for full license information.
 */

package com.microsoft.azure.iot.service.sdk;

import java.util.concurrent.CompletableFuture;

/**
* A receiver is a class abstracting the methods to receive messages,
* in order to use it for multiple endpoints, e.g. feedback endpoint,
* “alerts” (i.e. D2C).
*/
public abstract class Receiver
{
    protected abstract CompletableFuture<Void> openAsync();
    protected abstract CompletableFuture<Void> closeAsync();

    protected abstract CompletableFuture<FeedbackBatch> receiveAsync();
    protected abstract CompletableFuture<FeedbackBatch> receiveAsync(long timeoutSeconds);
}

File: csharp/service/Microsoft.Azure.Devices/Receiver.cs
// Copyright (c) Microsoft. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

namespace Microsoft.Azure.Devices
{
    using System;
    using System.Threading.Tasks;

    public abstract class Receiver<T>
    {
        /// <summary>
        /// Receive a message using the default timeout.
        /// </summary>
        /// <returns>The receive message or null if there was no message until the default timeout</returns>
        public abstract Task<T> ReceiveAsync();

        /// <summary>
        /// Receives a message
        /// </summary>
        /// <param name="timeout">The timeout for receiving a message</param>
        /// <returns>The receive message or null if there was no message until the specified timeout</returns>
        public abstract Task<T> ReceiveAsync(TimeSpan timeout);

        /// <summary>
        /// Deletes a received message from the queue
        /// </summary>
        public abstract Task CompleteAsync(T t);

        /// <summary>
        /// Puts a received message back into the queue
        /// </summary>
        public abstract Task AbandonAsync(T t);
    }
}
